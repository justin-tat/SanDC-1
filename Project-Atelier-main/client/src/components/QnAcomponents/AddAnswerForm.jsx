import React from 'react';
import axios from 'axios';
import AnswerPhotoUpload from './AnswerPhotoUpload.jsx';

class AddAnswerForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      answerBody: '',
      nickname: '',
      email: '',
      photos: [],
      isWarningShown: false
    };
    this.handleInputChange = this.handleInputChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
    this.handlePhotos = this.handlePhotos.bind(this);
  }

  handleInputChange(event) {
    const target = event.target;
    const value = target.value;
    const name = target.name;
    this.setState({
      [name]: value
    });
  }

  handlePhotos(photos) {
    console.log('triggered handle photos');
    this.setState({
      photos: photos
    });
  }

  handleSubmit(event) {
    event.preventDefault();
    var questionId = this.props.questionId;
    var body = this.state.answerBody;
    var nickname = this.state.nickname;
    var email = this.state.email;
    var photos = this.state.photos;
    var validationResult = this.handleValidation(body, nickname, email);
    var productId = this.props.productId;

    if (validationResult) {
      this.setState({
        isWarningShown: false
      });
      this.props.addNewAnswer(questionId, body, nickname, email, photos, productId);
      this.props.closeAnswer();


    } else {
      this.setState({
        isWarningShown: true
      });
    }
  }

  handleValidation(answer, nickname, email) {
    if (answer.length === 0) {
      return false;
    } else if (nickname.length === 0) {
      return false;
    } else if (!(/^\w+([\.-]?\w+)*@\w+([\.-]?\w+)*(\.\w{2,3})+$/.test(email))) {
      return false;
    } else {
      return true;
    }
  }

  render() {
    let warning;
    if (this.state.isWarningShown) {
      warning = <div className='qna-add-answer-form-warning-message'>You didn't filled all the fields properly</div>;
    } else {
      warning = <div></div>;
    }
    return (
      <div className='qna-add-answer-form'>
        <div className='qna-add-answer-form-wrapper'>
          <div className='qna-add-answer-form-subtitle'>Submit your answer</div>
          <div className='qna-add-answer-form-product-name'>{this.props.productName}: {this.props.question_body}</div>
          <form>
            <label>
              <div className='qna-add-answer-form-your-answer'>
              Your answer*
              </div>
              <input
                className='qna-add-answer-form-input-answer'
                name='answerBody'
                type='text'
                maxLength='1000'
                value={this.state.answerBody}
                onChange={this.handleInputChange}
              />
            </label>
            <br />
            <label>
              <div className='qna-add-answer-form-your-nickname'>
              What's your nickname?*
              </div>
              <input
                className='qna-add-answer-form-input-nickname'
                name='nickname'
                type='text'
                maxLength='60'
                placeholder='Example: jack543'
                value={this.state.nickname}
                onChange={this.handleInputChange}
              />
              <br />
              <div className='qna-add-answer-form-warning-1'>
              For privacy reasons, do not use your full name or email address
              </div>
            </label>
            <br />
            <label>
              <div className='qna-add-answer-form-your-email'>
              Your email?*
              </div>
              <input
                className='qna-add-answer-form-input-email'
                name='email'
                type='text'
                maxLength='60'
                placeholder='Example: jack@email.com'
                value={this.state.email}
                onChange={this.handleInputChange}
              />
              <br />
              <div className='qna-add-answer-form-warning-2'>
             For authentication reasons, you will not be emailed
              </div>
            </label>
            <br />
            <label>
              <br />
              <AnswerPhotoUpload handlePhotos={this.handlePhotos} />
            </label>
            {warning}
            <input type="submit" className='qna-add-answer-form-submit' value="Submit" onClick = {this.handleSubmit} />
            <button type='button' className='qna-add-answer-form-close' onClick={this.props.closeAnswer}>Close without adding</button>
          </form>
        </div>
      </div>
    );
  }
}

export default AddAnswerForm;