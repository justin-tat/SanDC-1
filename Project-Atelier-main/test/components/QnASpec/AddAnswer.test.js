import Enzyme from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';

Enzyme.configure({ adapter: new Adapter() });

import React from 'react';
import { shallow, mount, render, ShallowWrapper } from 'enzyme';
//example data
import exampleQuestions from '../../../example/questions.js';

import AddAnswerForm from '../../../client/src/components/QnAcomponents/AddAnswerForm.jsx';
import AnswerPhoto from '../../../client/src/components/QnAcomponents/AnswerPhotoUpload.jsx';

describe('Add new answer form', function() {

  test('click on submit button without filling the form', () => {
    const formEventMocked = { preventDefault: jest.fn() };
    const alertSpy = window.alert;
    window.alert = () => {}; // provide an empty implementation for window.alert

    const wrapper = shallow(<AddAnswerForm />);
    const spy1 = jest.spyOn(wrapper.instance(), 'handleInputChange');
    const spy2 = jest.spyOn(wrapper.instance(), 'handleSubmit');
    const spy3 = jest.spyOn(wrapper.instance(), 'handleValidation');
    //wrapper.find('[type="submit"]').simulate('submit');
    expect(wrapper.find('form')).toHaveLength(1);
    //wrapper.find('form').simulate('submit', formEventMocked);
    wrapper.find('[type="submit"]').simulate('click', formEventMocked);
    expect(formEventMocked.preventDefault).toBeCalledTimes(1);
    expect(spy3).toBeCalledTimes(1);
    expect(spy1).not.toHaveBeenCalled();

  });

  test('click on submit button when form not filled properly', () => {
    const formEventMocked = { preventDefault: jest.fn() };
    const alertSpy = window.alert;
    window.alert = () => {};

    const wrapper = shallow(<AddAnswerForm/>);
    const spy1 = jest.spyOn(wrapper.instance(), 'handleInputChange');
    const spy2 = jest.spyOn(wrapper.instance(), 'handleSubmit');
    const spy3 = jest.spyOn(wrapper.instance(), 'handleValidation');

    const state = {
      answerBody: 'This is an answer from tests',
      nickname: 'gandalf',
      email: ''
    };
    wrapper.setState(state);
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('form')).toHaveLength(1);

    wrapper.find('[type="submit"]').simulate('click', formEventMocked);

    expect(formEventMocked.preventDefault).toBeCalledTimes(1);
    expect(spy3).toBeCalledTimes(1);

  });

  test('allows to type in the form fields', () => {
    const formEventMocked = { preventDefault: jest.fn() };
    const alertSpy = window.alert;
    window.alert = () => {}; // provide an empty implementation for window.alert

    const wrapper = mount(<AddAnswerForm />);

    const answer = wrapper.find({name: 'answerBody'});
    answer.props().value = 'bar';

    expect(answer.props().value).toBeTruthy();

  });

  it('should render without throwing an error', function() {
    expect(shallow(<AddAnswerForm />).contains( <div className="qna-add-answer-form-subtitle">Submit your answer</div>)).toBe(true);
  });

  it('should be selectable by class', function() {
    expect(shallow(<AddAnswerForm />).is('.qna-add-answer-form')).toBe(true);
  });

  it('should mount in a full DOM', function() {
    //console.log(shallow(<AddAnswerForm />).debug());
    expect(shallow(<AddAnswerForm />).find('.qna-add-answer-form').length).toBe(1);
  });

  it('should render to static HTML', function() {
    var text = 'Submit your answer: Your answer*What\'s your nickname?*For privacy reasons, do not use your full name or email addressYour email?*For authentication reasons, you will not be emailedClose without adding';
    expect(render(<AddAnswerForm />).text()).toEqual(text);
  });

  test('handle validation should return false without input', () => {
    const component = new AddAnswerForm;
    expect(component.handleValidation('', '', '')).toEqual(false);
  });

  test('handle validation should return false without answer', () => {
    const component = new AddAnswerForm;
    expect(component.handleValidation('', 'aragorn', 'kings@dunedain.com')).toEqual(false);
  });

  test('handle validation should return false without nickname', () => {
    const component = new AddAnswerForm;
    expect(component.handleValidation('for frodo', '', 'kings@dunedain.com')).toEqual(false);
  });

  test('handle validation should return false without email', () => {
    const component = new AddAnswerForm;
    expect(component.handleValidation('for frodo', 'bill the pony', '')).toEqual(false);
  });

  test('handle validation should return false with email in wrong format', () => {
    const component = new AddAnswerForm;
    expect(component.handleValidation('for frodo', 'bill the pony', 'dummyemail')).toEqual(false);
  });

  test('handle validation should return true with proper arguments', () => {
    const component = new AddAnswerForm;
    expect(component.handleValidation('for frodo', 'bill the pony', 'ponies@example.com')).toEqual(true);
  });

  test('adding new answer invoked passed functions', () => {

    const formEventMocked = { preventDefault: jest.fn() };

    const passedMock1 = jest.fn();
    const passedMock2 = jest.fn();
    const wrapper = shallow(<AddAnswerForm addNewAnswer={passedMock1} closeAnswer={passedMock2} />);
    const spy1 = jest.spyOn(wrapper.instance(), 'handleInputChange');
    const spy2 = jest.spyOn(wrapper.instance(), 'handleSubmit');
    const spy3 = jest.spyOn(wrapper.instance(), 'handleValidation');

    const state = {
      answerBody: 'This is a answerfrom tests',
      nickname: 'gandalf',
      email: 'example@email.com'
    };
    wrapper.setState(state);
    expect(wrapper).toMatchSnapshot();

    wrapper.find('[type="submit"]').simulate('click', formEventMocked);
    expect(spy3).toBeCalledTimes(1);
    expect(passedMock1).toBeCalledTimes(1);
    expect(passedMock2).toBeCalledTimes(1);

  });
});

describe('Uploading photos in answers', function() {

  it('should be selectable by class', function() {
    expect(shallow(<AnswerPhoto />).is('.qna-answer-photo-upload')).toBe(true);
  });

  it('should mount in a full DOM', function() {
    expect(mount(<AnswerPhoto />).find('.qna-answer-photo-upload').length).toBe(1);
  });

  it('should render to static HTML', function() {
    var text = '';
    expect(render(<AnswerPhoto />).text()).toEqual(text);
  });
});