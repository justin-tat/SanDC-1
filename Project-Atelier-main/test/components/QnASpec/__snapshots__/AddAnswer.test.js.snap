// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add new answer form adding new answer invoked passed functions 1`] = `ShallowWrapper {}`;

exports[`Add new answer form click on submit button when form not filled properly 1`] = `ShallowWrapper {}`;
